src/Middleware/RequestThrottling/src/RequestThrottlingEventSource.cs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Text;
+using System.Threading;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.RequestThrottling
+{
+    internal sealed class RequestThrottlingEventSource : EventSource
+    {
+        public static readonly RequestThrottlingEventSource Log = new RequestThrottlingEventSource();
+        private static QueueFrame _cachedNonTimerResult = new QueueFrame
+        {
+            _parent = Log
+        };
+
+        private PollingCounter _rejectedRequestsCounter;   // incrementing polling counter
+        private PollingCounter _queueLengthCounter;        // polling counter
+        private EventCounter _queueDuration;     // event counter average
+
+        internal long _rejectedRequests;
+        internal int _queueLength;
+
+    
+        internal RequestThrottlingEventSource()
+            : base("Microsoft.AspNetCore.RequestThrottling")
+        {
+        }
+
+        // Used for testing
+        internal RequestThrottlingEventSource(string eventSourceName)
+            : base(eventSourceName)
+        {
+        }
+
+        [Event(1, Level = EventLevel.Warning)]
+        public void RequestRejected()
+        {
+            Interlocked.Increment(ref _rejectedRequests);
+            WriteEvent(1);
+        }
+
+        [NonEvent]
+        public QueueFrame QueueTimer()
+        {
+            Interlocked.Increment(ref _queueLength);
+
+            if (IsEnabled())
+            {
+                return new QueueFrame
+                {
+                    _timer = ValueStopwatch.StartNew(),
+                    _parent = this
+                };
+            }
+
+            return _cachedNonTimerResult;
+        }
+
+        internal struct QueueFrame : IDisposable
```

Comment:
This pattern is interesting for creating your own event source. Is this the natural way? Did you find docs/ info on how to do this?

Reply:
There aren't really good patterns here yet. This seems reasonable. It's pretty common to want to do "duration" counters like this (with an `IDiposable` that lets you wrap a lexical scope representing the operation). SignalR uses a different pattern of:

```csharp
public ValueStopwatch ConnectionStart(string connectionId);
public void ConnectionStop(string connectionId, ValueStopwatch timer)
```

Which is basically the same behavior but requires the caller to track it. In that case though, SignalR doesn't have a lexical scope that defines the lifetime of a connection.

Reply:
There's also 
```C#
public async Task TimeWaitInQueue(Task<bool> waitInQueueTask)
```
Creates the stopwatch, awaits the task, writes the results. Feels bad from a layered architecture standpoint.